### PR TITLE
Whitelist webpush endpoints

### PR DESCRIPTION
**Pull request by bwindels**
through an optional `allowed_endpoints` configuration key

I added support for globs because some browser vendor endpoints seem to contain sharding identifiers in the subdomain.

### CODE REVIEW

File: sygnal/webpushpushkin.py [hunk 1]
```diff
@@ -104,26 +113,42 @@ def __init__(self, name, sygnal, config):
...
+        endpoint_domain = urlparse(endpoint).netloc
+        if self.allowed_endpoints:
+            match = next(
+                (
+                    regex
+                    for regex in self.allowed_endpoints
+                    if regex.fullmatch(endpoint_domain)
+                ),
+                None,
+            )
```
Comment (clokep) [inline sygnal/webpushpushkin.py:139]: This is pretty difficult to read, I think you want something more like:

```python
allowed = any(regex.fullmatch(endpoint_domain) for regex in self.allowed_endpoints)
```

**Review by clokep — CHANGES_REQUESTED** [review]
Looks good overall.

**Comment by clokep** [comment]
Does it make sense to include some info on Google/Microsoft/Mozilla here so people know what a reasonable value is?

File: sygnal/webpushpushkin.py (hunk 1 above)
Comment (bwindels) [reply sygnal/webpushpushkin.py:139]: that is way better indeed, only came up with that monstrosity after looking online how to to do a `find`

**Comment by bwindels** [comment]
> Does it make sense to include some info on Google/Microsoft/Mozilla here so people know what a reasonable value is?

Added this to the docs :+1:

**Comment by bwindels** [comment]
How would I fix this error?:
```
sygnal/webpushpushkin.py:108: error: Incompatible types in assignment (expression has type "None", variable has type "List[Any]")  [assignment]
```

**Comment by clokep** [comment]
> How would I fix this error?:
> 
> ```
> sygnal/webpushpushkin.py:108: error: Incompatible types in assignment (expression has type "None", variable has type "List[Any]")  [assignment]
> ```

When you define `self.allowed_endpoints` you need to add a type to it. I usually do this as:

```python
self.foo = None  # type: Optional[List[str]]
if blah:
    self.foo = list(...)
```

You'll also need a `from typing import List, Optional` at the top.

File: changelog.d/182.feature [hunk 2]
```diff
@@ -0,0 +1 @@
+Add 'allowed_endpoints' configuration option for webpush pushkins to whitelist the allowed endpoints.
```
Comment (clokep) [inline changelog.d/182.feature:1]:
```suggestion
Add 'allowed_endpoints' configuration option for limiting the endpoints that WebPush pushkins will contact.
```

**Review by clokep — COMMENTED** [review]
Looks good! I think my other comment will fix the mypy issue.

**Review by clokep — APPROVED** [review]
LGTM!

**Pull request merged**